mlos_bench/mlos_bench/tests/environments/local_env_test.py

```diff
@@ -0,0 +1,68 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Unit tests for mock benchmark environment.
+"""
+from datetime import datetime, timedelta
+
+from mlos_bench.environments.local.local_env import LocalEnv
+from mlos_bench.services.config_persistence import ConfigPersistenceService
+from mlos_bench.services.local.local_exec import LocalExecService
+from mlos_bench.tunables.tunable_groups import TunableGroups
+
+
+def test_local_env(tunable_groups: TunableGroups) -> None:
+    """
+    Produce benchmark and telemetry data in a local script and read it.
+    """
+    ts1 = datetime.utcnow()
+    ts1 -= timedelta(microseconds=ts1.microsecond)
+    ts2 = ts1 + timedelta(minutes=1)
+
+    time_str1 = ts1.strftime("%Y-%m-%d %H:%M:%S")
+    time_str2 = ts2.strftime("%Y-%m-%d %H:%M:%S")
+
+    local_env = LocalEnv(
+        name="Test Local Env",
+        config={
+            "run": [
+                "echo 'metric,value' > output.csv",
+                "echo 'latency,10' >> output.csv",
+                "echo 'throughput,66' >> output.csv",
+                "echo 'score,0.9' >> output.csv",
+                "echo '-------------------'",  # This output does not go anywhere
+                "echo 'timestamp,metric,value' > telemetry.csv",
+                f"echo '\"{time_str1}\",cpu_load,0.65' >> telemetry.csv",
+                f"echo '\"{time_str1}\",mem_usage,10240' >> telemetry.csv",
+                f"echo '\"{time_str2}\",cpu_load,0.8' >> telemetry.csv",
+                f"echo '\"{time_str2}\",mem_usage,20480' >> telemetry.csv",
+            ],
+            "read_results_file": "output.csv",
+            "read_telemetry_file": "telemetry.csv",
+        },
+        tunables=tunable_groups,
+        service=LocalExecService(parent=ConfigPersistenceService()),
+    )
```

Comment:
Regarding the failures on Windows, I'd made a comment about that here:
https://github.com/microsoft/MLOS/pull/461#discussion_r1277930027

Reply:
Basically, the `echo 'metric,value' > output.csv` in Windows results in `metric,value ` (trailing space), and `pd.read_csv()` gets a column name of `value ` instead of `value`.

The fix I'd done there originally, was to special case that test and rename the columns to strip the trailing spaces on Windows like so:

```python
if sys.platform == 'win32':
    data.rename(str.rstrip, axis='columns', inplace=True)
```

I think I'd proposed doing that directly in the `local_exec` Env, but we decided to wait on it.

Could do that now and then remove the special case in that other test.

Thoughts?

Reply:
@bpkroth yes, I guess we have to strip the spaces when we read the data. I'll add your code to the PR.

That leads us to another problem: the lines
```python
"echo 'timestamp,metric,value' > telemetry.csv",
f"echo '\"{time_str1}\",cpu_load,0.65' >> telemetry.csv",
f"echo '\"{time_str1}\",mem_usage,10240' >> telemetry.csv",
f"echo '\"{time_str2}\",cpu_load,0.8' >> telemetry.csv",
f"echo '\"{time_str2}\",mem_usage,20480' >> telemetry.csv",
```
currently produce the following string:
```
timestamp,metric,value 
"\"2023-08-10 20:18:10\",cpu_load,0.65" 
"\"2023-08-10 20:18:10\",mem_usage,10240" 
"\"2023-08-10 20:19:10\",cpu_load,0.8" 
"\"2023-08-10 20:19:10\",mem_usage,20480" 
```
(with trailing spaces, of course)
do you have any idea how to fix that?
note that the timestamp can be like `2023-08-07 14:59:12,593000` - i.e., with a comma inside

Reply:
I think you can just drop the quotes.

Reply:
Oh wait, the timestamp includes a comma?

Reply:
and
```python
"echo 'timestamp,metric,value' > telemetry.csv",
f"echo '{time_str1},cpu_load,0.65' >> telemetry.csv",
f"echo '{time_str1},mem_usage,10240' >> telemetry.csv",
f"echo '{time_str2},cpu_load,0.8' >> telemetry.csv",
f"echo '{time_str2},mem_usage,20480' >> telemetry.csv",
```
gives
```csv
timestamp,metric,value 
"2023-08-10 20:28:37,cpu_load,0.65" 
"2023-08-10 20:28:37,mem_usage,10240" 
"2023-08-10 20:29:37,cpu_load,0.8" 
"2023-08-10 20:29:37,mem_usage,20480" 
```

Reply:
@bpkroth I agree, all those Windows shenanigans are probably not relevant to the real applications. Still, I just want to come up with a simple and human-readable unit test that works in both Linux and Windows

Reply:
Try this:

```
"echo timestamp,metric,value > telemetry.csv",
f"echo {time_str1},cpu_load,0.65 >> telemetry.csv",
f"echo {time_str1},mem_usage,10240 >> telemetry.csv",
f"echo {time_str2},cpu_load,0.8 >> telemetry.csv",
f"echo {time_str2},mem_usage,20480 >> telemetry.csv",
```

(no quotes)

Works locally at least, though I haven't tested it in the Python stack yet.

Reply:
wee! it works! Thank you!